### PR TITLE
Document providers and plugins

### PR DESCRIPTION
**Pull request by lunaris**
This commit stubs out developer documentation on the architecture and implementation of plugins and, more specifically, resource providers. Some sections are still incomplete but this is still more than we have today and hopefully serves as a good starting point.

### CODE REVIEW

**Comment by lunaris** [comment]
Relevant changes/new pages:

* https://pulumi-developer-docs--17159.org.readthedocs.build/17159/docs/architecture/providers.html
* https://pulumi-developer-docs--17159.org.readthedocs.build/17159/docs/architecture/plugins.html

Some links may be broken as I've split up a bunch of changes into smaller PRs

File: docs/architecture/plugins.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  Code](https://www.pulumi.com/docs/using-pulumi/crossguard/) product.
+* *Converter plugins* support the conversion of existing infrastructure as code
+  (e.g. [Terraform](https://github.com/pulumi/pulumi-converter-terraform)) to
+  Pulumi programs.
+* *Tool plugins* allow integrating Pulumi with arbitrary tools.
+
+A plugin's binary is typically named `pulumi-<kind>-<name>`, where `<kind>` is
+one of `resource`, `language`, `analyzer`, `converter`, or `tool`, and `<name>`
+is a unique name for the plugin. For example, the AWS resource provider plugin
+is named `pulumi-resource-aws`.
```
Comment (Frassle) [inline docs/architecture/plugins.md:28]: We should probably make a section about shimless here.

File: docs/architecture/providers.md [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+  library provides a high-level API for writing a provider in Go, without having
+  to worry about low-level gRPC details.
+* Each language SDK provides a `provider` package that offers low-level
+  primitives for writing a provider in that language (e.g.
+  <gh-file:pulumi#sdk/nodejs/provider> for NodeJS,
+  <gh-file:pulumi#sdk/python/lib/pulumi/provider> for Python, and
+  <gh-file:pulumi#sdk/go/pulumi/provider> for Go). These packages are in varying
+  states of completeness and neatness and are generally only used for building
+  [multi-language components (MLCs)](mlcs).
+
```
Comment (Frassle) [inline docs/architecture/providers.md:59]: Dynamic providers?

File: docs/architecture/providers.md [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+typically occurs:
+
+* When generating an SDK (e.g. using a `pulumi package add` command), we need to
+  boot up a provider and parameterize it using only information from the
+  command-line invocation. In this case, the parameter is a string array
+  representing the command-line arguments (`args`).
+* When interacting with a provider as part of program execution, the parameter
+  is *embedded in the SDK*, so as to free the program author from having to know
+  whether a provider is parameterized or not. In this case, the parameter is a
+  provider-specific, base-64-encoded bytestring (`value`). This is intended to
```
Comment (Frassle) [inline docs/architecture/providers.md:255]: Technically the base64 is a feature of the SDK encoding. The wire protocol is just byte[]

**Review by Frassle — APPROVED** [review]

**Pull request merged**